### PR TITLE
Kernel state is treated separately from other arguments

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #16246 Handle stack correctly&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13780147/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #16177 Op-calling API can handle state&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13742889/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#16178 Kernel state is treated separately from other arguments**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13742928/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #16180 Remove state from schema and calling API&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13744071/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #16247 Move stack.h to ATen/core&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13774189/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #16282 Kernel gets Stack* instead of ArrayRef<IValue>&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13792335/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #16283 Add a test case calling caffe2 layer_norm from caffe2 executor but through the c10 dispatcher&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13792591/)

Before, kernel state was just a normal operator argument.
This now changes and it is treated separately, which is not part of the operator signature.
In the next diff, operator state will be removed from the schema and calling API entirely and be a thing only the kernel knows about.

Differential Revision: [D13742928](https://our.internmc.facebook.com/intern/diff/D13742928/)

### CODE REVIEW

**Review by Ritzing — APPROVED** [review]

**Comment by dzhulgakov** [comment]
It might be much easier to review if it was merged with the next PR (https://github.com/pytorch/pytorch/pull/16180). They modify the same places (one adds State arg and the next deletes it). So it's much easier to comprehend that way.

**Comment by ezyang** [comment]
^--- what @dzhulgakov said

File: aten/src/ATen/core/dispatch/OpSchema.h [hunk 1]
```diff
@@ -10,7 +10,12 @@
 
 namespace c10 {
 
-using KernelFunction = IValue(ArrayRef<IValue>);
+class KernelState {
```
Comment (ezyang) [inline aten/src/ATen/core/dispatch/OpSchema.h]: This needs documentation

File: caffe2/core/operator_c10wrapper.h [hunk 2]
```diff
@@ -28,6 +28,9 @@ using extract_type_t = typename ParameterDef::type;
  * TODO: Figure out a better way to handle output parameters
  */
 
+// TODO This wrapper shouldn't have to handle state since it's only calling into c10
+class NoState final : public c10::KernelState {};
```
Comment (ezyang) [inline caffe2/core/operator_c10wrapper.h]: Wouldn't it make more sense to represent the NoState case as a nullptr?

Comment (smessmer) [reply caffe2/core/operator_c10wrapper.h]: NoState is gone now

**Pull request closed without merging**